keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1107 @@
+    coordination.k8s.io/compatibility-version: "1.29"
+  name: sample-controller-0001A
+  namespace: kube-system
+spec:
+  holderIdentity: sample-controller-0001A
+  leaseDurationSeconds: 10
+  renewTime: "2023-12-05T02:33:08.685777Z"
+```
+
+A component identity lease announces candidacy for leadership by including
+`coordination.k8s.io/can-lead-leases` it is identity lease. If the lease
+expires, the component is considered unavailable for leader election purposes.
+
+### Coordinated Election Controller
+
+A Coordinated Election Controller will reconcile component Leader Leases
+(primary resource) and Identity Leases (secondary resource, changes trigger
+reconciliation of related leader leases).
+
+Coordinated Election Controller reconciliation loop:
+
+- If no leader lease exists for a components:
+  - Elect leader from candidates by preparing a freshly renewed lease with:
+    - `spec.holderIdentity` set to the identity of the elected leader
+    - `coordination.k8s.io/elected-by: leader-election-controller` (to make
+      lease types easy to disambiguate)
+- If there is a better candidate than current leader:
+  - Sets `coordination.k8s.io/end-of-term: true` on the leader lease, signaling
+    that the leader should stop renewing the lease and yield leadership
+
+```mermaid
+flowchart TD
+   A[Reconcile] --> |Process Leader Lease| B
+   B{Lease Status?} --> |Better Leader Exists| D
+   B --> |Expired/Missing| E
+   D[End Lease Term]
+   E[Elect Leader]
+```
+
+Example of a lease created by Coordinated Election Controller:
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  annotations:
+    coordination.k8s.io/elected-by: coordinated-election-controller
+  name: sample-controller
+  namespace: kube-system
+spec:
+  holderIdentity: controller-a
+  leaseDurationSeconds: 10
+  leaseTransitions: 0
+  renewTime: "2023-12-05T18:58:31.295467Z"
+```
+
+The Coordinated Election Controller will run in the kube-apiserver.
+
+The apiserver runs very few controllers, and they are not elected, but instead
+all run concurrently in HA configurations.  Because of this, controller in the
+apiserver must make careful use of concurrency control primitives to ensure
+multiple instances collaborate, not fight. This is discussed in depth in the
+below design details.
+
+### Coordinated Lease Lock
+
+A new `resourceLock` type of `coordinatedleases`, and `CoordinatedLeaseLock`
+implementation of `resourcelock.Interface` will be added to client-go that:
+
+- Creates Identity Lease when ready to be Leader
+	- Renews identity lease periodically
+- Watches Leader Lease, waiting to be elected leader by the Coordinated Election Controller
+- When it becomes leader:
+  - Perform role of active component instance
+  - Renew leader lease periodically
+  - Stop renewing if lease is marked `coordination.k8s.io/end-of-term: true`
+- If leader lease expires:
+  - Shutdown (yielding leadership) and restart as a candidate component instance
```

Comment:
is this about shutting down and restarting a component instance as in e.g. a kube-* component restarting and obtaining a new PID when the component leader lease expires?
if yes, how?

i guess my confusion comes from that fact that this is under client-go.

Reply:
The exit() happens in the app code:
- https://github.com/kubernetes/kubernetes/blob/741fd84d1570caecc533152d3c6c9da6f5de8c4e/cmd/kube-scheduler/app/server.go#L239
- https://github.com/kubernetes/kubernetes/blob/741fd84d1570caecc533152d3c6c9da6f5de8c4e/cmd/kube-controller-manager/app/controllermanager.go#L303

Users of client-go don't necessarily need to follow this approach, but it's what we've used for the core controller managers..